server/src/audit/audit.guard.ts

```diff
@@ -0,0 +1,48 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { AuditService } from 'src/models/audit/audit.service';
+import { AUDITABLE_KEY } from './audit.decorator';
+
+@Injectable()
+export class AuditGuard implements CanActivate {
+  constructor(
+    private reflector: Reflector,
+    private auditService: AuditService,
+  ) {}
+
+  canActivate(context: ExecutionContext): boolean {
+    const isAuditable = this.reflector.getAllAndOverride<boolean>(
+      AUDITABLE_KEY,
+      [context.getHandler(), context.getClass()],
+    );
+
+    if (isAuditable) {
+      const { user, body, query, params } = context.switchToHttp().getRequest();
+
+      if (!user) {
+        throw new UnauthorizedException(
+          'Cannot determine user role as user is undefined',
+        );
+      }
+
+      this.auditService.createAudit({
+        endpointMethod: context.getHandler().name,
+        payload: JSON.stringify(body),
+        query: JSON.stringify(query),
+        params: JSON.stringify(params),
+        createdBy: {
+          connect: {
+            id: user.id,
+          },
+        },
+      });
+    }
```

Comment:
`nit`: I would still add a `.catch` just in case this promises fails out and use a logger to write a warning to logs